library/machine_provision.py

```diff
@@ -4,83 +4,91 @@
-
-    parameters['netboot_enabled'] = True
-
-    data = json.dumps(parameters)
-
-    r = requests.put(url, headers=headers, data=data)
-
-    if r.status_code != 200:
-        raise ProvisionerError('Error PUT {}, HTTP {} {}'.format(url,
-                         r.status_code, r.reason))
-    return r.json()
-
-def get_preseed_by_name(url, token, preseed_name):
-    """ Look up preseed by name """
-    headers = {'Authorization': token}
-    url = urljoin(url, "/api/v1/preseed?show_all=true")
-    r = requests.get(url, headers=headers)
-    if r.status_code != 200:
-        raise ProvisionerError('Error fetching {}, HTTP {} {}'.format(url,
-                         r.status_code, r.reason))
-    for preseed in r.json():
-        if preseed['name'] == preseed_name:
-            del preseed['content'] # we don't need it, and it's really big
-            return preseed
-
-    raise ProvisionerError('Error no preseed found with name "{}"'.
-            format(preseed_name))
-
-def get_image_by_description(url, token, image_type, description, arch):
-    """ Look up image by description """
-    headers = {'Authorization': token}
-    url = urljoin(url, "/api/v1/image?show_all=true")
-    r = requests.get(url, headers=headers)
-    if r.status_code != 200:
-        raise ProvisionerError('Error fetching {}, HTTP {} {}'.format(url,
-                         r.status_code, r.reason))
-    found_image = True
-    for image in r.json():
-        if (image['description'] == description and
-            image['type'] == image_type and
-            image['arch'] == arch):
-            return image
-    msg = "Error finding image of type '{}' and description '{}'".format(
-        image_type, description)
-    raise ProvisionerError(msg)
+from library.common import ProvisionerError
+from library.common import get_machine_by_name
+from library.image import ImageUploader
+from library.preseed import PreseedUploader
+
+
+class MachineProvisioner(object):
+    def __init__(self, mrp_url, mrp_token, machine_name):
+        self.mrp_url = mrp_url
+        self.mrp_token = mrp_token
+        self.machine_name = machine_name
+        self.machine_id = None
+        self.headers = {'Authorization': self.mrp_token}
+        self.validate_name()
+
+    def validate_name(self):
+        try:
+            machine = get_machine_by_name(
+                self.mrp_token, self.mrp_url, self.machine_name)
+            self.machine_id = machine['id']
+        except ProvisionerError as e:
+            print(e)
+
+    def machine_provision(self):
+        """ enables netboot on the machine and pxe boots it """
+        url = urljoin(
+            self.mrp_url, "/api/v1/machine/{}/state".format(self.machine_id))
+
+        data = json.dumps({'state': 'provision'})
+
+        r = requests.post(url, headers=self.headers, data=data)
+
+        if r.status_code not in [200, 202]:
+            raise ProvisionerError('Error PUTing {}, HTTP {} {}'.format(url,
```

Comment:
PUT or POST?
This is repeated over and over and is prone to copy-pasta, needs to be lifted into a URL handling library.